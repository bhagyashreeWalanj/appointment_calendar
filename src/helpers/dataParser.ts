import { eachHourOfInterval, format, addDays, startOfWeek} from'date-fns'

import {status} from '../actions/constants'

export const initData =(slotsData: any):any =>{
    const segregatedSlots: any = {};

    slotsData.calendar.forEach((element : any) => {
        const slotDateTime = new Date(element.date_time);
        const year = slotDateTime.getFullYear();
        const month = slotDateTime.getMonth();
        const day = slotDateTime.getDate();
        
        if(!segregatedSlots[year]){segregatedSlots[year] = {};};
        if(!segregatedSlots[year][month]){segregatedSlots[year][month] = {}};

        if(!segregatedSlots[year][month][day]){
            segregatedSlots[year][month][day] = [slotDateTime.getHours()]
        }else{
            segregatedSlots[year][month][day].push(slotDateTime.getHours())
        }
    });

    return segregatedSlots;
}

export const compareTwoDates = (dt1: Date, dt2: Date) => {
    const date1 = (new Date(dt1)).setHours(0,0,0,0);
    const date2 = (new Date(dt2)).setHours(0,0,0,0);
    return date1 > date2;
};

/*
slotTimes: number[] = [6, 12, 19]
selectedDateTime: Date = Fri Jul 15 2022 12:49:38 GMT+0200
*/
export const getAllSlotsWithStatus = (slotsInHrsArray: number[], selectedDateTime: Date) => {
   const allocatedSlots = getAllocatedSlotsForDate(selectedDateTime);
    const todayDt = new Date();
    const currentHour = todayDt > selectedDateTime ? todayDt.getHours() : -1;
    const allTimeSlots: any = [];
    const hourDiffArray = getTimeDifferByHour();
      hourDiffArray.forEach((date, index) => {
        const dt = new Date(date)
      //  const nextHour = addHours(dt, 1);
        const hour = dt.getHours();
        let formattedTime = formatHourTime(dt)
        allTimeSlots.push({
            hour:hour,
            timeSlot: formattedTime,
            status: ((slotsInHrsArray && slotsInHrsArray.includes(hour)) || hour<=currentHour)
            ? status.NOT_AVAILABLE : (allocatedSlots.includes(hour) ?  status.BOOKED : status.AVAILABLE)
        });
      });
    return allTimeSlots;
};

export const getAllocatedSlotsForDate= (selectedDateTime: Date) : number[] =>{
    const allocatedSlotsString = localStorage.getItem(selectedDateTime.getDate().toString());
    return allocatedSlotsString?allocatedSlotsString.split(",").map(Number):[]
}

/*
getTimeDifferByHour Response : 
(24) [Tue Feb 01 2022 00:00:00 GMT+0100 (Central European Standard Time), 
    Tue Feb 01 2022 01:00:00 GMT+0100 (Central European Standard Time), 
    
    ...
     Tue Feb 01 2022 22:00:00 GMT+0100 (Central European Standard Time) 
     Tue Feb 01 2022 23:00:00 GMT+0100 (Central European Standard Time) {}length: 24[[Prototype]]: Array(0)

*/
const getTimeDifferByHour =()=>{
    var result = eachHourOfInterval({
        start: new Date(2022, 1, 1, 0),
        end: new Date(2022, 1, 1, 23)
      })
      return result;
}

const formatHourTime = (dt: Date) => {
   return  `${format(dt, 'HH:00')}`
};

export const getWeekNames=(today: Date)=>{
    const startDate = startOfWeek(today, { weekStartsOn: 0 })
    const dayFormat = 'eee'
    const weeks: any = []

    for (let i = 0; i < 7; i++) {
      weeks.push(format(addDays(startDate, i), dayFormat))
    }
    return weeks;
}