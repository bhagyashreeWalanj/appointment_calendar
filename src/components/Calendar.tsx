import { useState, useEffect } from 'react'
import style from '../styles/Calendar.module.scss'

import {
  IoIosArrowDropleftCircle,
  IoIosArrowDroprightCircle,
} from 'react-icons/io'
import {
  add,
  eachDayOfInterval,
  endOfMonth,
  format,
  getDay,
  isEqual,
  isToday,
  parse,
  startOfToday,
} from 'date-fns'

import { compareTwoDates, getWeekNames } from '../helpers/dataParser'

function classNames(...classes: any) {
  return classes.filter(Boolean).join(' ')
}
let colStartClasses = [
  '',
  style.col_start_2,
  style.col_start_3,
  style.col_start_4,
  style.col_start_5,
  style.col_start_6,
  style.col_start_7,
]

interface IReactCalendar {
  selectedDayFn(day: Date): void
}

const Calendar = ({ selectedDayFn }: IReactCalendar) => {
  let today = startOfToday()
  let [selectedDay, setSelectedDay] = useState(today)
  const [weekDays, setWeekDays] = useState([])
  let [currentMonth, setCurrentMonth] = useState(format(today, 'MMM-yyyy'))
  let firstDayCurrentMonth = parse(currentMonth, 'MMM-yyyy', new Date())

  let days = eachDayOfInterval({
    start: firstDayCurrentMonth,
    end: endOfMonth(firstDayCurrentMonth),
  })

  function previousMonth() {
    let firstDayNextMonth = add(firstDayCurrentMonth, { months: -1 })
    setCurrentMonth(format(firstDayNextMonth, 'MMM-yyyy'))
  }

  function nextMonth() {
    let firstDayNextMonth = add(firstDayCurrentMonth, { months: 1 })
    setCurrentMonth(format(firstDayNextMonth, 'MMM-yyyy'))
  }

  useEffect(() => {
    const weeks = getWeekNames(today)
    setWeekDays(weeks)
  }, [])

  const handleOnSelectDate = (day: Date) => {
    setSelectedDay(day)
    selectedDayFn(day)
  }
  document.getElementById('acc-close')?.click()
  return (
    <div>
      <div className={style.calendar}>
        <div className={style.month}>
          <label className={style.nav} onClick={previousMonth}>
            <IoIosArrowDropleftCircle />
          </label>
          <div>
            <span className={style.year}>
              {' '}
              {format(firstDayCurrentMonth, 'MMMM yyyy')}
            </span>
          </div>
          <label className={style.nav} onClick={nextMonth}>
            <IoIosArrowDroprightCircle />
          </label>
        </div>
        <div className={style.days}>
          {weekDays.map((week) => {
            return <span key={week}>{week}</span>
          })}
        </div>
        <div className={style.dates}>
          {days.map((day, dayIdx) => (
            <div
              key={day.toString()}
              className={classNames(
                dayIdx === 0 && colStartClasses[getDay(day)],
              )}
            >
              <button
                type="button"
                id={`calendarDateBtn_${dayIdx}`}
                onClick={() => handleOnSelectDate(day)}
                disabled={compareTwoDates(today, day) ? true : false}
                className={classNames(
                  !isEqual(day, selectedDay) &&
                    isToday(day) &&
                    style.nonSelectedCurrentDay &&
                    style.currentDay,
                  compareTwoDates(today, day) && style.disable_days,

                  isEqual(day, selectedDay) && isToday(day) && style.currentDay,
                  isEqual(day, selectedDay) &&
                    !isToday(day) &&
                    style.isNotToday,
                )}
              >
                <time
                  dateTime={format(day, 'yyyy-MM-dd')}
                  id={format(day, 'd')}
                >
                  {format(day, 'd')}
                </time>
              </button>
            </div>
          ))}
        </div>
      </div>
    </div>
  )
}

export default Calendar
