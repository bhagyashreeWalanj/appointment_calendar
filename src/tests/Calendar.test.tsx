import React from 'react'
import { shallow } from 'enzyme'
import Calendar from '../components/Calendar'
import { getDate } from 'date-fns'
import { compareTwoDates } from '../helpers/dataParser'
import {
  add,
  addDays,
  eachDayOfInterval,
  endOfMonth,
  format,
  getDay,
  isEqual,
  isToday,
  parse,
  startOfToday,
  startOfWeek,
} from 'date-fns'
describe('Testing Calendar Component', () => {
  const selectedDayFn = jest.fn()
  const timeSlots = {
    mentor: {
      name: 'Max Mustermann',
      time_zone: '-03:00',
    },
    calendar: [
      {
        date_time: '2022-07-24 00:37:02 +0200',
      },
      {
        date_time: '2022-07-24 03:31:50 +0200',
      },
      {
        date_time: '2022-07-24 12:13:56 +0200',
      },
      {
        date_time: '2022-07-25 20:49:10 +0200',
      },
    ],
  }

  const wrapper = shallow(<Calendar selectedDayFn={selectedDayFn} />)
  let todayDate = getDate(new Date())

  it('should display to Home Component', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should display current Day', () => {
    const container = wrapper.find('.currentDay')
    expect(container.contains(todayDate.toString())).toBe(true)
  })

  it('should disable past dates', () => {
    const isPastDate = compareTwoDates(
      new Date(),
      new Date('2022-07-02T22:00:00.000Z'),
    )
    const disableDayContainer = wrapper.find(`.disable_days time`).at(9)

    expect(disableDayContainer.html().includes('10')).toBe(true)
    expect(isPastDate).toBe(true)
  })

  it('should not allow to select disable date and call function', () => {
    const disableDayContainer = wrapper.find(`.disable_days time`).at(9)
    expect(disableDayContainer.html().includes('10')).toBe(true)
    expect(selectedDayFn).toBeCalledTimes(0)
  })
  it('should display get selected Date Slot List', () => {
    wrapper.find('#calendarDateBtn_17').simulate('click')
    expect(selectedDayFn).toBeCalledTimes(1)
  })
})
