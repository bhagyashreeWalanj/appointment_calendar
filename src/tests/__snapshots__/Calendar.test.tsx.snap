// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Calendar Component should display to Home Component 1`] = `
<div>
  <div
    className="calendar"
  >
    <div
      className="month"
    >
      <a
        className="nav"
        href="#"
        onClick={[Function]}
      >
        <IoIosArrowDropleftCircle />
      </a>
      <div>
        <span
          className="year"
        >
           
          July 2022
        </span>
      </div>
      <a
        className="nav"
        href="#"
        onClick={[Function]}
      >
        <IoIosArrowDroprightCircle />
      </a>
    </div>
    <div
      className="days"
    />
    <div
      className="dates"
    >
      <div
        className="col_start_6"
        key="Fri Jul 01 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_0"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-01"
            id="1"
          >
            1
          </time>
        </button>
      </div>
      <div
        className=""
        key="Sat Jul 02 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_1"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-02"
            id="2"
          >
            2
          </time>
        </button>
      </div>
      <div
        className=""
        key="Sun Jul 03 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_2"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-03"
            id="3"
          >
            3
          </time>
        </button>
      </div>
      <div
        className=""
        key="Mon Jul 04 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_3"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-04"
            id="4"
          >
            4
          </time>
        </button>
      </div>
      <div
        className=""
        key="Tue Jul 05 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_4"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-05"
            id="5"
          >
            5
          </time>
        </button>
      </div>
      <div
        className=""
        key="Wed Jul 06 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_5"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-06"
            id="6"
          >
            6
          </time>
        </button>
      </div>
      <div
        className=""
        key="Thu Jul 07 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_6"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-07"
            id="7"
          >
            7
          </time>
        </button>
      </div>
      <div
        className=""
        key="Fri Jul 08 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_7"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-08"
            id="8"
          >
            8
          </time>
        </button>
      </div>
      <div
        className=""
        key="Sat Jul 09 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_8"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-09"
            id="9"
          >
            9
          </time>
        </button>
      </div>
      <div
        className=""
        key="Sun Jul 10 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_9"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-10"
            id="10"
          >
            10
          </time>
        </button>
      </div>
      <div
        className=""
        key="Mon Jul 11 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_10"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-11"
            id="11"
          >
            11
          </time>
        </button>
      </div>
      <div
        className=""
        key="Tue Jul 12 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_11"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-12"
            id="12"
          >
            12
          </time>
        </button>
      </div>
      <div
        className=""
        key="Wed Jul 13 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_12"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-13"
            id="13"
          >
            13
          </time>
        </button>
      </div>
      <div
        className=""
        key="Thu Jul 14 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_13"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-14"
            id="14"
          >
            14
          </time>
        </button>
      </div>
      <div
        className=""
        key="Fri Jul 15 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_14"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-15"
            id="15"
          >
            15
          </time>
        </button>
      </div>
      <div
        className=""
        key="Sat Jul 16 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="disable_days"
          disabled={true}
          id="calendarDateBtn_15"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-16"
            id="16"
          >
            16
          </time>
        </button>
      </div>
      <div
        className=""
        key="Sun Jul 17 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className="currentDay"
          disabled={false}
          id="calendarDateBtn_16"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-17"
            id="17"
          >
            17
          </time>
        </button>
      </div>
      <div
        className=""
        key="Mon Jul 18 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_17"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-18"
            id="18"
          >
            18
          </time>
        </button>
      </div>
      <div
        className=""
        key="Tue Jul 19 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_18"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-19"
            id="19"
          >
            19
          </time>
        </button>
      </div>
      <div
        className=""
        key="Wed Jul 20 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_19"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-20"
            id="20"
          >
            20
          </time>
        </button>
      </div>
      <div
        className=""
        key="Thu Jul 21 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_20"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-21"
            id="21"
          >
            21
          </time>
        </button>
      </div>
      <div
        className=""
        key="Fri Jul 22 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_21"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-22"
            id="22"
          >
            22
          </time>
        </button>
      </div>
      <div
        className=""
        key="Sat Jul 23 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_22"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-23"
            id="23"
          >
            23
          </time>
        </button>
      </div>
      <div
        className=""
        key="Sun Jul 24 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_23"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-24"
            id="24"
          >
            24
          </time>
        </button>
      </div>
      <div
        className=""
        key="Mon Jul 25 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_24"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-25"
            id="25"
          >
            25
          </time>
        </button>
      </div>
      <div
        className=""
        key="Tue Jul 26 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_25"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-26"
            id="26"
          >
            26
          </time>
        </button>
      </div>
      <div
        className=""
        key="Wed Jul 27 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_26"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-27"
            id="27"
          >
            27
          </time>
        </button>
      </div>
      <div
        className=""
        key="Thu Jul 28 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_27"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-28"
            id="28"
          >
            28
          </time>
        </button>
      </div>
      <div
        className=""
        key="Fri Jul 29 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_28"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-29"
            id="29"
          >
            29
          </time>
        </button>
      </div>
      <div
        className=""
        key="Sat Jul 30 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_29"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-30"
            id="30"
          >
            30
          </time>
        </button>
      </div>
      <div
        className=""
        key="Sun Jul 31 2022 00:00:00 GMT+0200 (Central European Summer Time)"
      >
        <button
          className=""
          disabled={false}
          id="calendarDateBtn_30"
          onClick={[Function]}
          type="button"
        >
          <time
            dateTime="2022-07-31"
            id="31"
          >
            31
          </time>
        </button>
      </div>
    </div>
  </div>
</div>
`;
